src/MongoDB.Driver/Encryption/ExplicitEncryptionLibMongoCryptController.cs

```diff
@@ -0,0 +1,285 @@
+                }
+                else
+                {
+                    throw new ArgumentException("Either keyId or alternateKeyName must be provided.");
+                }
+
+                using (context)
+                {
+                    var wrappedBytes = await ProcessStatesAsync(context, databaseName: null, cancellationToken).ConfigureAwait(false);
+                    return UnwrapEncryptedValue(wrappedBytes);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new MongoEncryptionException(ex);
+            }
+        }
+
+        // private methods
+        private IKmsKeyId GetKmsKeyId(string kmsProvider, IReadOnlyList<string> alternateKeyNames, BsonDocument masterKey)
+        {
+            IEnumerable<byte[]> wrappedAlternateKeyNamesBytes = null;
+            if (alternateKeyNames != null)
+            {
+                wrappedAlternateKeyNamesBytes = alternateKeyNames.Select(GetWrappedAlternateKeyNameBytes);
+            }
+
+            switch (kmsProvider)
+            {
+                case "aws":
+                    var customerMasterKey = masterKey["key"].ToString();
+                    var region = masterKey["region"].ToString();
+                    return wrappedAlternateKeyNamesBytes != null
+                        ? new AwsKeyId(customerMasterKey, region, wrappedAlternateKeyNamesBytes)
+                        : new AwsKeyId(customerMasterKey, region);
+                case "local":
+                    return wrappedAlternateKeyNamesBytes != null ? new LocalKeyId(wrappedAlternateKeyNamesBytes) : new LocalKeyId();
+                default:
+                    throw new ArgumentException($"Invalid kmsProvider {kmsProvider}.");
+            }
+        }
+
+        private byte[] GetWrappedAlternateKeyNameBytes(string value)
+        {
+            return
+               !string.IsNullOrWhiteSpace(value)
+                   ? new BsonDocument("keyAltName", value).ToBson()
+                   : null;
+        }
+
+        private byte[] GetWrappedValueBytes(BsonValue value)
+        {
+            var wrappedValue = new BsonDocument("v", value);
+            var writerSettings = BsonBinaryWriterSettings.Defaults.Clone();
+            writerSettings.GuidRepresentation = GuidRepresentation.Unspecified;
+            return wrappedValue.ToBson(serializer: BsonValueSerializer.Instance, writerSettings: writerSettings);
+        }
+
+        private BsonValue UnwrapDecryptedValue(byte[] wrappedBytes)
+        {
+            var wrappedDocument = new RawBsonDocument(wrappedBytes);
+            return wrappedDocument["v"];
+        }
+
+        private BsonBinaryData UnwrapEncryptedValue(byte[] encryptedWrappedBytes)
+        {
+            var wrappedDocument = new RawBsonDocument(encryptedWrappedBytes);
+            return wrappedDocument["v"].AsBsonBinaryData;
+        }
+
+        private BsonBinaryData UnwrapKeyId(RawBsonDocument wrappedKeyDocument)
```

Comment:
It's not really "unwrap" logic, named in this way just to follow the same naming as we use for other fields when we prepare the method result. 
Another name can be just `GetKeyId`, but we already have a very similar name `GetKmsKeyId`. Can be discussed.

Reply:
Change return type to `Guid`.

Reply:
`UnwrapKeyId` name is OK.